google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcConversions.java

```diff
@@ -165,26 +192,279 @@ private BucketInfo bucketInfoDecode(Bucket from) {
     // TODO(frankyn): Add decoder for iamConfig
-    // TODO(frankyn): Add Etag when support is avialable
+    // TODO(frankyn): Add Etag when support is available
+    return to.build();
+  }
+
+  private Bucket bucketInfoEncode(BucketInfo from) {
+    Bucket.Builder to = Bucket.newBuilder();
+    to.setName(from.getName());
+    to.setBucketId(from.getGeneratedId());
+    if (from.getRetentionPeriodDuration() != null) {
+      Bucket.RetentionPolicy.Builder retentionPolicyBuilder = to.getRetentionPolicyBuilder();
+      ifNonNull(
+          from.getRetentionPeriodDuration(),
+          durationMillisCodec::encode,
+          retentionPolicyBuilder::setRetentionPeriod);
+      ifNonNull(from.retentionPolicyIsLocked(), retentionPolicyBuilder::setIsLocked);
+      if (from.retentionPolicyIsLocked() == Boolean.TRUE) {
+        ifNonNull(
+            from.getRetentionEffectiveTimeOffsetDateTime(),
+            timestampCodec::encode,
+            retentionPolicyBuilder::setEffectiveTime);
+      }
+      to.setRetentionPolicy(retentionPolicyBuilder.build());
+    }
+    ifNonNull(from.getLocation(), to::setLocation);
+    ifNonNull(from.getLocationType(), to::setLocationType);
+    ifNonNull(from.getMetageneration(), to::setMetageneration);
+    Bucket.Billing.Builder billingBuilder = Billing.newBuilder();
+    ifNonNull(from.requesterPays(), billingBuilder::setRequesterPays);
+    to.setBilling(billingBuilder.build());
+    ifNonNull(from.getCreateTimeOffsetDateTime(), timestampCodec::encode, to::setCreateTime);
+    ifNonNull(from.getUpdateTimeOffsetDateTime(), timestampCodec::encode, to::setUpdateTime);
+    Bucket.Encryption.Builder encryptionBuilder = Bucket.Encryption.newBuilder();
+    ifNonNull(from.getDefaultKmsKeyName(), encryptionBuilder::setDefaultKmsKey);
+    to.setEncryption(encryptionBuilder.build());
+    Bucket.Website.Builder websiteBuilder = Bucket.Website.newBuilder();
+    ifNonNull(from.getIndexPage(), websiteBuilder::setMainPageSuffix);
+    ifNonNull(from.getNotFoundPage(), websiteBuilder::setNotFoundPage);
+    to.setWebsite(websiteBuilder.build());
+    ifNonNull(from.getRpo(), Rpo::toString, to::setRpo);
+    ifNonNull(from.getStorageClass(), StorageClass::toString, to::setStorageClass);
+    Bucket.Versioning.Builder versioningBuilder = Bucket.Versioning.newBuilder();
+    ifNonNull(from.versioningEnabled(), versioningBuilder::setEnabled);
+    to.setVersioning(versioningBuilder.build());
+    ifNonNull(from.getDefaultEventBasedHold(), to::setDefaultEventBasedHold);
+    ifNonNull(from.getLabels(), to::putAllLabels);
+    // preserve mapping to deprecated property
+    Bucket.Lifecycle.Builder lifecycleBuilder = Bucket.Lifecycle.newBuilder();
+    ifNonNull(
+        from.getLifecycleRules(),
+        Utils.toImmutableListOf(lifecycleRule()::encode),
+        lifecycleBuilder::addAllRule);
+    /*
+    // Corruption occurs by having deleteRule() values added
```

Comment:
@BenWhitehead deleteRule() corrupts the policy and I commented it out. I can do a follow-up PR to better handle this case as this PR is long. WDYT?

Reply:
SGTM